### PR TITLE
More of object

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/logic/object/update/specialpowerupdatemodule.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+public:
+    virtual ~SpecialPowerUpdateModule() override;
+
+    virtual void CRC_Snapshot(Xfer *xfer) override;
+    virtual void Xfer_Snapshot(Xfer *xfer) override;
+    virtual void Load_Post_Process() override;
+
+    virtual bool Does_Special_Power_Update_Pass_Science_Test() const override;
+    virtual ScienceType Get_Extra_Required_Science() const override;
+};
```
Comment (xezon) [inline src/game/logic/object/update/specialpowerupdatemodule.h]: eol

File: src/game/logic/object/object.cpp [hunk 2]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+                drawable->Clear_Model_Condition_State(MODELCONDITION_SNOW);
+            } else if (int_param == 2) {
+                drawable->Set_Model_Condition_State(MODELCONDITION_SNOW);
+            }
+        }
+
+        bool ambient_enabled_exists;
+        bool ambient_enabled = properties->Get_Bool(g_objectSoundAmbientEnabled, &ambient_enabled_exists);
+        DynamicAudioEventInfo *info = nullptr;
+        bool set_info = false;
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: Perhaps call it info_ok

File: src/game/logic/object/object.cpp [hunk 3]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+            }
+        }
+
+        if (!exists || !string_param.Is_Empty()) {
+            bool_param = properties->Get_Bool(g_objectSoundAmbientCustomized, &exists);
+
+            if (exists && bool_param) {
+                if (info == nullptr) {
+                    const AudioEventInfo *base_info = drawable->Get_Base_Sound_Ambient_Info();
+                    captainslog_dbgassert(base_info != nullptr, "getBaseSoundAmbientInfo() return NULL");
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: Get_Base_Sound_Ambient_Info

File: src/game/logic/object/object.cpp [hunk 4]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+    SimpleObjectIterator *iter = g_thePartitionManager->Iterate_Objects_In_Range(Get_Position(),
+        Get_Geometry_Info().Get_Bounding_Circle_Radius() + 25.0f,
+        FROM_CENTER_2D,
+        filters,
+        ITER_SORTED_NEAR_TO_FAR);
+    MemoryPoolObjectHolder holder(iter);
+
+    Object *found = nullptr;
+
+    for (Object *o = iter->First(); o != nullptr; o = iter->Next()) {
```
Comment (xezon) [inline src/game/logic/object/object.cpp:3806]: obj

File: src/game/logic/object/object.cpp [hunk 5]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+                Set_Team(team);
+            }
+        } else {
+            captainslog_dbgassert(
+                false, "Object original team (%s) could not be found or created!", m_originalTeamName.Str());
+        }
+    }
+}
+
+bool Object::Check_And_Detonate_Booby_Trap(Object *obj)
```
Comment (xezon) [inline src/game/logic/object/object.cpp:3786]: nearby_obj

File: src/game/logic/object/object.cpp [hunk 6]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+    PartitionFilter *filters[] = { &filter1, &filter2, nullptr };
+
+    SimpleObjectIterator *iter = g_thePartitionManager->Iterate_Objects_In_Range(Get_Position(),
+        Get_Geometry_Info().Get_Bounding_Circle_Radius() + 25.0f,
+        FROM_CENTER_2D,
+        filters,
+        ITER_SORTED_NEAR_TO_FAR);
+    MemoryPoolObjectHolder holder(iter);
+
+    Object *found = nullptr;
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: bomb_obj

File: src/game/logic/object/object.cpp [hunk 7]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+        }
+    }
+
+    if (found == nullptr) {
+        return false;
+    }
+
+    static NameKeyType key_StickyBombUpdate = g_theNameKeyGenerator->Name_To_Key("StickyBombUpdate");
+
+    StickyBombUpdate *update = static_cast<StickyBombUpdate *>(found->Find_Update_Module(key_StickyBombUpdate));
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: bomb_update

File: src/game/logic/object/object.cpp [hunk 8]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+    if (contain != nullptr) {
+        contain->Iterate_Contained(Local_Is_Hero, &hero, false);
+    }
+
+    return hero || Is_KindOf(KINDOF_HERO);
+}
+
+unsigned int Object::Get_Disabled_Until(DisabledType type) const
+{
+    if (type == 0xFFFF) {
```
Comment (xezon) [inline src/game/logic/object/object.cpp:3913]: `DISABLED_TYPE_INVALID = 0xFFFF` ?

File: src/game/logic/object/object.cpp [hunk 9]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+}
+
+bool Object::Affected_By_Upgrade(const UpgradeTemplate *upgrade) const
+{
+    BitFlags<128> mask = Get_Controlling_Player()->Get_Upgrades_Completed();
+    mask.Set(Get_Object_Upgrade_Mask());
+    mask.Set(upgrade->Get_Upgrade_Mask());
+
+    for (BehaviorModule **module = m_allModules; *module != nullptr; module++) {
+        UpgradeModuleInterface *upgrade = (*module)->Get_Upgrade();
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: This variable masks the other upgrade variable.

upgrade_module

File: src/game/logic/object/object.cpp [hunk 10]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+
+    return false;
+}
+
+void Object::Remove_Upgrade(const UpgradeTemplate *upgrade)
+{
+    m_objectUpgradesCompleted.Clear(upgrade->Get_Upgrade_Mask());
+
+    for (BehaviorModule **module = m_allModules; *module != nullptr; module++) {
+        UpgradeModuleInterface *interface = (*module)->Get_Upgrade();
```
Comment (xezon) [inline src/game/logic/object/object.cpp:4047]: upgrade_module

File: src/game/logic/object/object.cpp [hunk 11]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+        g_theRadar->Remove_Object(this);
+    }
+
+    Drawable *drawable = Get_Drawable();
+
+    if (drawable != nullptr) {
+        drawable->Set_Terrain_Decal_Fade_Target(0.0f, -0.03f);
+    }
+
+    Object *obj = g_theGameLogic->Find_Object_By_ID(Get_Producer_ID());
```
Comment (xezon) [inline src/game/logic/object/object.cpp:4082]: producer_obj

File: src/game/logic/object/object.cpp [hunk 12]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+            g_theRadar->Try_Event(RADAR_EVENT_UNIT_LOST, Get_Position());
+        }
+    }
+
+    if (Get_Controlling_Player() != nullptr) {
+        g_theInGameUI->Remove_Idle_Worker(this, Get_Controlling_Player()->Get_Player_Index());
+    }
+
+    if (Get_Status(OBJECT_STATUS_RECONSTRUCTING)) {
+        Object *producer = g_theGameLogic->Find_Object_By_ID(Get_Producer_ID());
```
Comment (xezon) [inline src/game/logic/object/object.cpp:4112]: We already have the producer_obj from line 4082

File: src/game/logic/object/object.cpp [hunk 13]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+            RebuildHoleBehaviorInterface *hole =
+                RebuildHoleBehavior::Get_Rebuild_Hole_Behavior_Interface_From_Object(producer);
+            captainslog_dbgassert(hole != nullptr, "Object::onDie() -  No Rebuild Hole Behavior interface on hole");
+
+            if (hole != nullptr) {
+                hole->Start_Rebuild_Process(Get_Template(), Get_ID());
+            }
+
+            for (Object *o = g_theGameLogic->Get_First_Object(); o != nullptr; o = o->Get_Next_Object()) {
+                AIUpdateInterface *update = Get_AI_Update_Interface();
```
Comment (xezon) [inline src/game/logic/object/object.cpp:4123]: ai_update

File: src/game/logic/object/object.cpp [hunk 14]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+        if (producer != nullptr) {
+            RebuildHoleBehaviorInterface *hole =
+                RebuildHoleBehavior::Get_Rebuild_Hole_Behavior_Interface_From_Object(producer);
+            captainslog_dbgassert(hole != nullptr, "Object::onDie() -  No Rebuild Hole Behavior interface on hole");
+
+            if (hole != nullptr) {
+                hole->Start_Rebuild_Process(Get_Template(), Get_ID());
+            }
+
+            for (Object *o = g_theGameLogic->Get_First_Object(); o != nullptr; o = o->Get_Next_Object()) {
```
Comment (xezon) [inline src/game/logic/object/object.cpp:4122]: obj

File: src/game/logic/object/object.cpp [hunk 15]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+                if (update != nullptr) {
+                    update->Transfer_Attack(Get_ID(), producer->Get_ID());
+                }
+            }
+        }
+    }
 }
 
-unsigned int Object::Get_Last_Shot_Fired_Frame() const
+void Object::Do_Special_Power(const SpecialPowerTemplate *t, unsigned int i, bool b)
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: const SpecialPowerTemplate *special_power
unsigned int options
bool force_usable

File: src/game/logic/object/object.cpp [hunk 16]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+        SpecialPowerModuleInterface *power = Get_Special_Power_Module(t);
+
+        if (power != nullptr) {
+            power->Do_Special_Power(i);
+        }
+    }
 }
 
-bool Object::Attempt_Healing_From_Sole_Benefactor(float f, const Object *obj, unsigned int i)
+void Object::Do_Special_Power_At_Object(const SpecialPowerTemplate *t, Object *obj, unsigned int i, bool b)
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: const SpecialPowerTemplate *special_power
unsigned int options
bool force_usable

File: src/game/logic/object/object.cpp [hunk 17]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
 
-void Object::Set_Effectively_Dead(bool dead) {}
+        if (power != nullptr) {
+            power->Do_Special_Power_At_Object(obj, i);
+        }
+    }
+}
 
-bool Object::Is_Hero() const
+void Object::Do_Special_Power_At_Location(const SpecialPowerTemplate *t, const Coord3D *pos, float f, unsigned int i, bool b)
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: const SpecialPowerTemplate *special_power
unsigned int options
bool force_usable

File: src/game/logic/object/object.cpp [hunk 18]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+        SpecialPowerModuleInterface *power = Get_Special_Power_Module(t);
+
+        if (power != nullptr) {
+            power->Do_Special_Power_At_Location(pos, f, i);
+        }
+    }
 }
 
-unsigned int Object::Get_Disabled_Until(DisabledType type) const
+void Object::Do_Special_Power_Using_Waypoints(const SpecialPowerTemplate *t, const Waypoint *wp, unsigned int i, bool b)
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: const SpecialPowerTemplate *special_power
const Waypoint *waypoint
unsigned int options
bool force_usable

File: src/game/logic/object/object.cpp [hunk 19]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+                    ProductionUpdateInterface *production = Get_Production_Update_Interface();
 
-void Object::Set_Layer(PathfindLayerEnum layer) {}
+                    if (production == nullptr && tmplate == nullptr) {
+                        goto l1;
+                    }
 
-void Object::Friend_Prepare_For_Map_Boundary_Adjust() {}
+                    production->Queue_Create_Unit(tmplate, production->Request_Unique_Unit_ID());
+                    break;
```
Comment (xezon) [inline src/game/logic/object/object.cpp:4197]: Make this a `return` in each switch case.

File: src/game/logic/object/object.cpp [hunk 20]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+        if (button != nullptr) {
+            switch (button->Get_Command()) {
+                case GUI_COMMAND_DOZER_CONSTRUCT:
+                case GUI_COMMAND_UNIT_BUILD: {
+                    const ThingTemplate *tmplate = *button->Get_Template();
+                    ProductionUpdateInterface *production = Get_Production_Update_Interface();
 
-void Object::Set_Layer(PathfindLayerEnum layer) {}
+                    if (production == nullptr && tmplate == nullptr) {
+                        goto l1;
```
Comment (xezon) [inline src/game/logic/object/object.cpp:4193]: Make this a `break` in each switch case.

File: src/game/logic/object/object.cpp [hunk 21]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+                    }
+
+                    update->AI_Hack_Internet(type);
+                    break;
+                }
+                case GUI_COMMAND_SWITCH_WEAPON:
+                    Set_Weapon_Lock(button->Get_Weapon_Slot(), LOCKED_LEVEL_2);
+                default: {
+                l1:
+                    captainslog_dbgassert(false,
```
Comment (xezon) [inline src/game/logic/object/object.cpp:4270]: Move this out of switch default. Remove the goto label.

File: src/game/logic/object/object.cpp [hunk 22]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+                case GUI_COMMAND_HACK_INTERNET: {
+                    if (update == nullptr) {
+                        goto l1;
+                    }
+
+                    update->AI_Hack_Internet(type);
+                    break;
+                }
+                case GUI_COMMAND_SWITCH_WEAPON:
+                    Set_Weapon_Lock(button->Get_Weapon_Slot(), LOCKED_LEVEL_2);
```
Comment (xezon) [inline src/game/logic/object/object.cpp:4265]: `break`, `return` missing.

File: src/game/logic/object/object.cpp [hunk 23]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+        AIUpdateInterface *update = Get_AI_Update_Interface();
+
+        if (button != nullptr) {
+            switch (button->Get_Command()) {
+                case GUI_COMMAND_STOP: {
+                    if (update != nullptr) {
+                        update->AI_Idle(type);
+                    }
+
+                    break;
```
Comment (xezon) [inline src/game/logic/object/object.cpp:4292]: Same as with function above. Get rid of `goto` by returning here and turning `goto` labels into breaks.

File: src/game/logic/object/object.cpp [hunk 24]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+{
+    if (!Is_Disabled()) {
+        AIUpdateInterface *update = Get_AI_Update_Interface();
+
+        if (button != nullptr) {
+            switch (button->Get_Command()) {
+                case GUI_COMMAND_DOZER_CONSTRUCT: {
+                    g_theBuildAssistant->Build_Object_Now(
+                        this, *button->Get_Template(), pos, 0.0f, Get_Controlling_Player());
+                    break;
```
Comment (xezon) [inline src/game/logic/object/object.cpp:4367]: Same as with function above. Get rid of `goto` by returning here and turning `goto` labels into breaks.

File: src/game/logic/object/object.cpp [hunk 25]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+        }
+    }
+}
 
 DockUpdateInterface *Object::Get_Dock_Update_Interface()
 {
+    for (BehaviorModule **module = m_allModules; *module != nullptr; module++) {
+        DockUpdateInterface *dock = (*module)->Get_Dock_Update_Interface();
+
+        if (dock != nullptr) {
```
Comment (xezon) [inline src/game/logic/object/object.cpp:4455]: Obsolete test. If it is null, return null.

File: src/game/logic/object/object.cpp [hunk 26]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+
     return nullptr;
 }
 
 ProjectileUpdateInterface *Object::Get_Projectile_Update_Interface() const
 {
+    for (BehaviorModule **module = m_allModules; *module != nullptr; module++) {
+        ProjectileUpdateInterface *projectile = (*module)->Get_Projectile_Update_Interface();
+
+        if (projectile != nullptr) {
```
Comment (xezon) [inline src/game/logic/object/object.cpp:4468]: Obsolete test. If it is null, return null.

File: src/game/logic/object/object.cpp [hunk 27]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+            }
+        }
+    }
+}
 
-void Object::Do_Command_Button_Using_Waypoints(const CommandButton *button, const Waypoint *wp, CommandSourceType type) {}
+void Object::Do_Command_Button_Using_Waypoints(const CommandButton *button, const Waypoint *wp, CommandSourceType type)
+{
+    if (!Is_Disabled() && button == nullptr) {
+        if ((button->Get_Options() & 0x400000) != 0) {
```
Comment (xezon) [inline src/game/logic/object/object.cpp:4433]: CAN_USE_WAYPOINTS

File: src/game/logic/object/object.cpp [hunk 28]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
     WhichTurretType type, const char *bone, Coord3D *pos, Matrix3D *tm) const
 {
-    return false;
+    if (Get_Drawable() == nullptr || Get_AI_Update_Interface() == nullptr) {
+        return false;
+    }
+
+    Coord3D rot;
+    Get_Drawable()->Get_Projectile_Launch_Offset(WEAPONSLOT_PRIMARY, 1, nullptr, type, &rot, nullptr);
+    Coord3D position;
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: bone_pos

File: src/game/logic/object/object.cpp [hunk 29]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+
+    if (Get_Drawable()->Get_Pristine_Bone_Positions(bone, 0, &position, nullptr, 1) != 1) {
+        return false;
+    }
+
+    float turret_angle;
+    Get_AI_Update_Interface()->Get_Turret_Rot_And_Pitch(type, &turret_angle, nullptr);
+
+    Matrix3D m(true);
+    m.Translate(Vector3(position.x, position.y, position.z));
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: bone_m

File: src/game/logic/object/object.cpp [hunk 30]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
 
 bool Object::Get_Single_Logical_Bone_Position_On_Turret(
     WhichTurretType type, const char *bone, Coord3D *pos, Matrix3D *tm) const
 {
-    return false;
+    if (Get_Drawable() == nullptr || Get_AI_Update_Interface() == nullptr) {
+        return false;
+    }
+
+    Coord3D rot;
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: launch_offset

File: src/game/logic/object/object.cpp [hunk 31]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+        return false;
+    }
+
+    float turret_angle;
+    Get_AI_Update_Interface()->Get_Turret_Rot_And_Pitch(type, &turret_angle, nullptr);
+
+    Matrix3D m(true);
+    m.Translate(Vector3(position.x, position.y, position.z));
+
+    Matrix3D m2(true);
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: turret_m

File: src/game/logic/object/object.cpp [hunk 32]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+    m2.Translate(rot.x, rot.y, rot.z);
+    m2.In_Place_Pre_Rotate_Z(turret_angle);
+
+    m2.Translate(-rot.x, -rot.y, -rot.z);
+
+    Matrix3D m3;
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: turret_bone_m

File: src/game/logic/object/object.cpp [hunk 33]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+    m3.Mul(m2, m);
+
+    Matrix3D m4;
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: world_turret_bone_m

File: src/game/logic/object/object.cpp [hunk 34]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+    Convert_Bone_Pos_To_World_Pos(nullptr, &m3, nullptr, &m4);
+    Vector3 translation = m4.Get_Translation();
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: world_bone_pos

File: src/game/logic/object/object.cpp [hunk 35]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+    }
+
+    if (tm != nullptr) {
+        *tm = m4;
+    }
+
+    return true;
 }
 
 int Object::Get_Multi_Logical_Bone_Position(const char *bone, int i, Coord3D *pos, Matrix3D *tm, bool b) const
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: const char *bone_name
int max_count
Coord3D *pos_array
Matrix3D *tm_array
bool world_space

File: src/game/logic/object/object.cpp [hunk 36]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+            if (pos != nullptr) {
+                p = &pos[j];
+            } else {
+                p = nullptr;
+            }
+
+            Matrix3D *m2;
+
+            if (tm != nullptr) {
+                m2 = &tm[j];
```
Comment (xezon) [inline src/game/logic/object/object.cpp:4614]: This looks as if it should be simply

```c++
Matrix3D *m2 = m;
Coord3D *p2 = p;
```

Or something is wrong here?

File: src/game/logic/object/object.cpp [hunk 37]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+            }
+
+            m_drawable->Convert_Bone_Pos_To_World_Pos(p2, m2, p, m);
+        }
+    }
+
+    return count;
 }
 
 bool Object::Can_Produce_Upgrade(const UpgradeTemplate *t)
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: upgrade

File: src/game/logic/object/object.cpp [hunk 38]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+        if (button != nullptr && button->Get_Upgrade_Template() != nullptr && button->Get_Upgrade_Template() == t) {
+            return true;
+        }
+    }
+
     return false;
 }
 
-void Object::Go_Invulnerable(unsigned int i) {}
+void Object::Go_Invulnerable(unsigned int i)
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: timer

File: src/game/logic/object/object.cpp [hunk 39]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+    RadarPriorityType priority = Get_Template()->Get_Radar_Priority();
+
+    if (priority == RADAR_PRIORITY_NOT_ON_RADAR) {
+        ContainModuleInterface *contain = Get_Contain();
+
+        if (contain && contain->Is_Garrisonable()) {
+            priority = RADAR_PRIORITY_UNIT;
+        }
+
+        if (Is_KindOf(KINDOF_CAPTURABLE)) {
```
Comment (xezon) [inline src/game/logic/object/object.cpp:4669]: else if

File: src/game/logic/object/object.cpp [hunk 40]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+    }
+
+    return priority;
 }
 
 ObjectID Object::Calculate_Countermeasure_To_Divert_To(const Object &obj)
 {
+    if (Get_AI_Update_Interface() != nullptr) {
+        for (BehaviorModule **module = obj.Get_All_Modules(); *module != nullptr; module++) {
+            CountermeasuresBehaviorInterface *interface = (*module)->Get_Countermeasures_Behavior_Interface();
```
Comment (xezon) [inline src/game/logic/object/object.cpp:4685]: countermeasures

File: src/game/logic/object/object.cpp [hunk 41]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+                    g_theBuildAssistant->Sell_Object(this);
+                    break;
+                }
+                case GUI_COMMAND_FIRE_WEAPON: {
+                    if (update == nullptr) {
+                        goto l1;
+                    }
 
-void Object::Update_Obj_Values_From_Map_Properties(Dict *properties) {}
+                    if ((button->Get_Options() & 7) != 0 || (button->Get_Options() & 0x20) != 0) {
```
Comment (xezon) [inline src/game/logic/object/object.cpp:4236]: Do we have any clue on what these bit fields are?

File: src/game/logic/object/object.cpp [hunk 42]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
-}
+    if (!Is_Disabled()) {
+        AIUpdateInterface *update = Get_AI_Update_Interface();
 
-void Object::Force_Refresh_Sub_Object_Upgrade_Status() {}
+        if (button != nullptr) {
+            switch (button->Get_Command()) {
+                case GUI_COMMAND_DOZER_CONSTRUCT:
+                case GUI_COMMAND_UNIT_BUILD: {
+                    const ThingTemplate *tmplate = *button->Get_Template();
```
Comment (xezon) [inline src/game/logic/object/object.cpp:4189]: thing

File: src/game/logic/object/object.cpp [hunk 43]
```diff
@@ -3428,150 +3434,1255 @@ void Object::Clear_Leech_Range_Mode_For_All_Weapons()
...
+{
+    Check_And_Detonate_Booby_Trap(nullptr);
+#ifdef GAME_DEBUG_STRUCTS
+    captainslog_dbgassert(!m_hasDiedAlready, "Object::onDie has been called multiple times. This is invalid.");
+    m_hasDiedAlready = true;
+#endif
+    bool source = damage->m_in.m_sourceID == Get_ID();
+
+    for (BehaviorModule **module = m_allModules; *module != nullptr; module++) {
+        DieModuleInterface *die = (*module)->Get_Die();
```
Comment (xezon) [inline src/game/logic/object/object.cpp:4065]: die_module

File: src/game/logic/object/update/specialpowerupdatemodule.h (hunk 1 above)
Comment (jonwil) [reply src/game/logic/object/update/specialpowerupdatemodule.h]: Fixed.

File: src/game/logic/object/object.cpp (hunk 2 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 3 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 6 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 7 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 9 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 11 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp:4082]: fixed.

File: src/game/logic/object/object.cpp (hunk 15 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 16 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 17 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 18 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 22 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp:4265]: fixed.

File: src/game/logic/object/object.cpp (hunk 25 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp:4455]: Not obsolete, if its nullptr we want to continue looping.

File: src/game/logic/object/object.cpp (hunk 26 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp:4468]: Not obsolete, if its nullptr we want to continue looping.

File: src/game/logic/object/object.cpp (hunk 28 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 29 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 30 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 31 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 32 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 33 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 34 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 36 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp:4614]: No this is what the original code does. Want to keep it that way to make comparing it to pseudo easier.

File: src/game/logic/object/object.cpp (hunk 38 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 41 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp:4236]: No.

File: src/game/logic/object/object.cpp (hunk 39 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp:4669]: Original has no else, want to keep it matching original.

File: src/game/logic/object/object.cpp (hunk 20 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp:4193]: I want to keep this matching pseudo to make it easier to spot any problems that might show up.

File: src/game/logic/object/object.cpp (hunk 21 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp:4270]: I want to keep this matching pseudo to make it easier to spot any problems that might show up.

File: src/game/logic/object/object.cpp (hunk 23 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp:4292]: I want to keep this matching pseudo to make it easier to spot any problems that might show up.

File: src/game/logic/object/object.cpp (hunk 24 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp:4367]: I want to keep this matching pseudo to make it easier to spot any problems that might show up.

File: src/game/logic/object/object.cpp (hunk 19 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp:4197]: I want to keep this matching pseudo to make it easier to spot any problems that might show up.

File: src/game/logic/object/object.cpp (hunk 37 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 35 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

**Comment by xezon** [comment]
Shell Map looks as usual
Demo Trap triggers and detonates ok
~~Helicopter sometimes does not lift off immediately after loading with passengers~~
Hero unit works
Salvage crates work
Healing works
Aircraft works
Unit upgrades work
~~Infantry Unit does not follow Barracks waypoint when leaving it after repair~~
Command Buttons work
Docking works
Special Powers work
Countermeasures work

**Review by xezon — APPROVED** [review]
Looks ok.

**Pull request merged**